Update PostbankPDFExtractor.java

Clemens82 opened the pull request:
Fixed issues with PostbankPDFExtractor and added compatibility for more postbank document types.

Considered Cases:
- "Rücknahme Investmentfonds" (Opposit of existing A"usgabe Investmentfonds")
- "Date" Section parsing (some, especially older documents have no "Schlusstag" but only a "Datum" section)
- "Schlusstag/-Zeit" in some documents has no "Auftragserteilung" behind the date. Existing pattern did not match because it required a space after the date
- Added "Gutschrift von Investmenterträgen" as Header for dividends (older documents used that heading).

Nirus2000 commented:
Hallo @Clemens82 
Bitte die PDF-Debug's für 
- "Rücknahme Investmentfonds"
- "Gutschrift von Investmenterträgen" 

und den TestCase dafür erstellen. Dieser fehlt noch in deinem PR.

Die TestCases findest du [hier](https://github.com/buchen/portfolio/blob/a9d3dfbaf32dd184b929f86e5d3e138873d129ad/name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/postbank/PostbankPDFExtractorTest.java).

Gruß
Alex

Clemens82 commented:
I added some test cases for new documents I added and also for issues I fixed with the previous version.

There's one open question left (see postbank_verkauf04.txt; testDividende04() ):
For dividends, the section "Anrechenbare Quellensteuer" is added to the final amount ("Ausmachender Betrag"). I am no expert of Quellensteuer and taxes, so I don't know why portfolio performance is doing it and if it's correct. I would not do it...

See this statement:
![grafik](https://user-images.githubusercontent.com/10797307/142399014-9264f34c-33ed-4cae-a497-73e6dd451c61.png)
After parsing, portfolio perfoamance comes up with final amount 19,25+0,29=19,54.

Any advice if this is correct (and I should adapt the unit test accordingly) or what else should be done here?

Nirus2000 commented on name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/PostbankPDFExtractor.java:
```diff
@@ -1,288 +1,325 @@
+                        .match("^Anrechenbare Quellensteuer [.,\\d]+ % .* [.,\\d]+ \\w{3} (?<quellenstanr>[.,\\d]+) (?<currency>\\w{3})$")
+                        .assign((t, v) -> addTax(type, t, v, "quellenstanr"))
+                        
+                        // Anrechenbare Quellensteuer pro Stück 0,0144878 EUR 0,29 EUR
```
Here is your fault. Delete this.

Nirus2000 commented:
Hallo @Clemens82
I think, that's correct.

Look closely at the document. The important words are "_Ausländiche Dividende_" and "_Inländische Dividende_".
There is €0.29 withholding tax on the foreign dividends.

Most double taxation agreements stipulate that a maximum of XY% of the withholding tax paid abroad is automatically creditable in Germany. The banks automatically offset in these cases.

For more informationen -> [here](https://wendl-koehler.de/rechtliches/erstattung-auslaendischer-quellensteuer/) <-

Clemens82 commented:
Thanks, then this PR should be done. Parser improved and 4 new tests added.

Nirus2000 commented:
Hello @Clemens82 
Plz check your PDF-Debugs. Some personal information like IBAN, Depotnummer, Auftragsnummer etc., and change it.
Alex

Nirus2000 commented on name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/PostbankPDFExtractor.java:
```diff
@@ -1,283 +1,325 @@
+
+                        // Is type --> "Verkauf" change from BUY to SELL
+                        .section("type").optional().match("Wertpapier Abrechnung (?<type>(Verkauf|R.cknahme Investmentfonds)).*").assign((t, v) -> {
+                            if (v.get("type").equals("Verkauf") || v.get("type").contains("cknahme Investmentfonds"))
```
Maybe change this to this?
`if (v.get("type").equals("Verkauf") || v.get("type").equals("Rücknahme Investmentfonds)"))`

Clemens82 replied:
I wanted to avoid the 'ü' because of potential encoding problems. That's why in the Regex there's also "R.cknahme".

Clemens82 commented:
Thanks for the hints. Documents should be completely anonymized now and I added "umlaute" as suggested.
